src/theory/strings/regexp_elim.cpp

```diff
@@ -0,0 +1,441 @@
+/*********************                                                        */
+/*! \file regexp_elim.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2018 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Implementation of techniques for eliminating regular expressions
+ **
+ **/
+
+#include "theory/strings/regexp_elim.h"
+
+#include "options/strings_options.h"
+#include "theory/strings/theory_strings_rewriter.h"
+
+using namespace CVC4;
+using namespace CVC4::kind;
+using namespace CVC4::theory;
+using namespace CVC4::theory::strings;
+
+RegExpElimination::RegExpElimination()
+{
+  d_zero = NodeManager::currentNM()->mkConst(Rational(0));
+  d_one = NodeManager::currentNM()->mkConst(Rational(1));
+  d_neg_one = NodeManager::currentNM()->mkConst(Rational(-1));
+}
+
+Node RegExpElimination::eliminate(Node atom)
+{
+  Assert(atom.getKind() == STRING_IN_REGEXP);
+  if (atom[1].getKind() == REGEXP_CONCAT)
+  {
+    return eliminateConcat(atom);
+  }
+  else if (atom[1].getKind() == REGEXP_STAR)
+  {
+    return eliminateStar(atom);
+  }
+  return Node::null();
+}
+
+Node RegExpElimination::eliminateConcat(Node atom)
+{
+  NodeManager* nm = NodeManager::currentNM();
+  Node x = atom[0];
+  Node lenx = nm->mkNode(STRING_LENGTH, x);
+  Node re = atom[1];
+  // memberships of the form x in re.++ * s1 * ... * sn *, where * are
+  // any number of repetitions (exact or indefinite) of re.allchar.
+  Trace("re-elim-debug") << "Try re concat with gaps " << atom << std::endl;
+  std::vector<Node> children;
+  TheoryStringsRewriter::getConcat(re, children);
+  bool success = true;
+  std::vector<Node> sep_children;
+  std::vector<unsigned> gap_minsize;
+  std::vector<bool> gap_exact;
+  // the first gap is initially strict zero
+  gap_minsize.push_back(0);
+  gap_exact.push_back(true);
+  unsigned nchildren = children.size();
+  for (unsigned i = 0; i < nchildren; i++)
+  {
+    Node c = children[i];
+    Trace("re-elim-debug") << "  " << c << std::endl;
+    success = false;
+    if (c.getKind() == STRING_TO_REGEXP)
```

Comment:
It seems like we should also be able to deal with `REGEXP_RANGE` here, no? Ranges would just add a constraint that if you extract that char, it matches the regexp range.

Reply:
This would definitely need to be a "aggressive" rewrite. We maybe could do this (although out of scope for this PR).